spec.emu

```diff
@@ -23,83 +23,121 @@ contributors: Krzysztof Kotowicz, Mike Samuel
-    1. Else,
-      1. Assert: _argCount_ &gt; 1.
-      1. Let _firstArg_ be _args_[0].
-      1. <ins>If IsCodeLike(_firstArg_) is *true*, set _firstArg_ to _firstArg_.[[HostDefinedCodeLike]]. Otherwise, set _assembledFromCodeLike_ to *false*.</ins>
-      1. Set _P_ to ? ToString(_firstArg_).
+    1. <ins>If IsCodeLike(_bodyArg_) is *true*, set _bodyArg_ to _bodyArg_.[[HostDefinedCodeLike]]. Otherwise, set _assembledFromCodeLike_ to *false*.</ins>
+    1. Let _bodyString_ be ? ToString(_bodyArg_).
+    1. Let _parameterStrings_ be a new empty List.
+    1. For each element _arg_ of _parameterArgs_, do
+      1. <ins>If IsCodeLike(_arg_) is *true*, set _arg_ to _arg_.[[HostDefinedCodeLike]]. Otherwise, set _assembledFromCodeLike_ to *false*.</ins>
+      1. Append ? ToString(_arg_) to _parameterStrings_.
+    1. Let _currentRealm_ be the current Realm Record.
+    1. <del>Perform ? <del>HostEnsureCanCompileStrings(_currentRealm_, _parameterStrings_, _bodyString_, *false*).</del>
+    1. Let _P_ be the empty String.
+    1. If _argCount_ > 0, then
+      1. Set _P_ to _parameterStrings_[0].
       1. Let _k_ be 1.
-      1. Repeat, while _k_ &lt; _argCount_ - 1,
-        1. Let _nextArg_ be _args_[_k_].
-        1. <ins>If IsCodeLike(_nextArg_) is *true*, set _nextArg_ to _nextArg_.[[HostDefinedCodeLike]]. Otherwise, set _assembledFromCodeLike_ to *false*.</ins>
-        1. Let _nextArgString_ be ? ToString(_nextArg_).
+      1. Repeat, while _k_ &lt; _argCount_,
+        1. Let _nextArgString_ be _parameterStrings_[_k_].
         1. Set _P_ to the string-concatenation of _P_, *","* (a comma), and _nextArgString_.
         1. Set _k_ to _k_ + 1.
-      1. Let _bodyArg_ be _args_[_k_].
-      1. <ins>If IsCodeLike(_bodyArg_) is *true*, set _bodyArg_ to _bodyArg_.[[HostDefinedCodeLike]]. Otherwise, set _assembledFromCodeLike_ to *false*.</ins>
-    1. Let _bodyString_ be the string-concatenation of 0x000A (LINE FEED), ? ToString(_bodyArg_), and 0x000A (LINE FEED).
-    1. Let _prefix_ be the prefix associated with _kind_ in <emu-xref href="#table-dynamic-function-sourcetext-prefixes"></emu-xref>.
-    1. Let _sourceString_ be the string-concatenation of _prefix_, *" anonymous("*, _P_, 0x000A (LINE FEED), *") {"*, _bodyString_, and *"}"*.
-    1. <ins>Set _sourceString_ to be ? HostValidateDynamicCode(_callerRealm_, _calleeRealm_, _sourceString_, _assembledFromCodeLike_, *"Function"*). </ins>
+    1. Let _bodyParseString_ be the string-concatenation of 0x000A (LINE FEED), _bodyString_, and 0x000A (LINE FEED).
+    1. Let _sourceString_ be the string-concatenation of _prefix_, *" anonymous("*, _P_, 0x000A (LINE FEED), *") {"*, _bodyParseString_, and *"}"*.
+    1. <ins>Set _sourceString_ to ? HostValidateDynamicCode(_currentRealm_, _parameterStrings_, _bodyString_, _sourceString_, *false*, _assembledFromCodeLike_, *"Function"*).</ins>
```

Comment:
Given we have to construct the _sourceString_ here anyway it seemed worthwhile to just move the call to HostEnsureCanCompileStrings/ HostValidateDynamicCode below the code that does so. This should have no behaviour changes.